docs/proposals/apiserver_watch.md

```diff
@@ -0,0 +1,132 @@
+## Abstract
+
+In the current system, all watch requests send to apiserver are in general
+redirected to etcd. This means that for every watch request to apiserver,
+apiserver opens a watch on etcd.
+
+The purpose of the proposal is to improve the overall performance of the system
+by solving the following problems:
+
+- having too many open watches on etcd
+- avoiding deserializing/converting the same objects multiple times in different
+watch results
+
+In the future, we would also like to add an indexing mechanism to the watch.
+Although Indexer is not part of this proposal, it is supposed to be compatible
+with it - in the future Indexer should be incorporated into the proposed new
+watch solution in apiserver without requiring any redesign.
+
+
+## High level design
+
+We are going to solve those problems by allowing many clients to watch the same
+storage in the apiserver, without being redirected to etcd.
+
+At the high level, apiserver will have a single watch open to etcd, watching all
+the objects (of a given type) without any filtering. The changes delivered from
+etcd will then be stored in a cache in apiserver. The introduced cache will have
+a limited capacity - whenever a new change comes from etcd and the cache is full,
+the oldest change will be removed to make place for the new change.
```

Comment:
Index would be against deserialized API objects though, not against
underlying etcd objects.  So that presupposes a cache of API objects above
etcd that you index, which implies you hold those API objects in memory.
There doesn't seem to be much value in converting the etcd object to the
internal version for the index and then throwing that away.

Or are you proposing indexing on raw etcd json, in which case we would be
converting the entire list every time we made a list call?

If we index above etcd we also are duplicating the lower level objects,
which has some memory overhead.

On Jul 15, 2015, at 5:01 PM, Daniel Smith <notifications@github.com> wrote:

In docs/proposals/apiserver_watch.md
<https://github.com/GoogleCloudPlatform/kubernetes/pull/10679#discussion_r34728704>
:

> +In the future, we would also like to add an indexing mechanism to the watch.
> +Although Indexer is not part of this proposal, it is supposed to be compatible
> +with it - in the future Indexer should be incorporated into the proposed new
> +watch solution in apiserver without requiring any redesign.
> +
> +
> +## High level design
> +
> +We are going to solve those problems by allowing many clients to watch the same
> +storage in the apiserver, without being redirected to etcd.
> +
> +At the high level, apiserver will have a single watch open to etcd, watching all
> +the objects (of a given type) without any filtering. The changes delivered from
> +etcd will then be stored in a cache in apiserver. The introduced cache will have
> +a limited capacity - whenever a new change comes from etcd and the cache is full,
> +the oldest change will be removed to make place for the new change.

It's necessary if you want your response to LIST to be reasonably fast. We
have to have that to index, so we might as well do it now.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/10679/files#r34728704>
.